dali/operators/reader/loader/video/video_file.h

```diff
@@ -0,0 +1,172 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_READER_LOADER_VIDEO_VIDEO_FILE_H_
+#define DALI_OPERATORS_READER_LOADER_VIDEO_VIDEO_FILE_H_
+
+extern "C" {
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+#include <libavutil/avutil.h>
+#include <libswscale/swscale.h>
+}
+
+#include "dali/core/common.h"
+
+#include <vector>
+#include <string>
+
+
+namespace dali {
+struct IndexEntry {
+  int last_keyframe_id;
+  int pts;
+  bool is_keyframe;
+  bool is_flush_frame;
+};
+
+struct AvState {
+  AVFormatContext *ctx_ = nullptr;
+  AVCodec *codec_ = nullptr;
+  AVCodecParameters *codec_params_ = nullptr;
+  AVCodecContext *codec_ctx_ = nullptr;
+  AVFrame *frame_ = nullptr;
+  AVPacket *packet_ = nullptr;
+  SwsContext  *sws_ctx_ = nullptr;
+  int stream_id_ = -1;
+
+  ~AvState() {
+    sws_freeContext(sws_ctx_);
+    avformat_close_input(&ctx_);
+    avformat_free_context(ctx_);
+    av_frame_free(&frame_);
+    av_packet_free(&packet_);
+    avcodec_free_context(&codec_ctx_);
+
+    ctx_ = nullptr;
+    codec_ = nullptr;
+    codec_params_ = nullptr;
+    codec_ctx_ = nullptr;
+    frame_ = nullptr;
+    packet_ = nullptr;
+    sws_ctx_ = nullptr;
+  }
+};
+
+/**
+ * @brief Object representing a video file. Allows access to frames and seeking.
+ * 
+ */
+class DLL_PUBLIC VideoFileCPU {
+ public:
+
+  /**
+   * @brief Construct a new VideoFileCPU object.
+   * 
+   * @param filename Path to a video file.
+   */
+  VideoFileCPU(const std::string &filename);
+
+  /**
+   * @brief Number of frames in the video
+   * 
+   * @return int64_t 
+   */
+  int64_t NumFrames() const {
+    return index_.size();
+  }
+
+  /**
+   * @brief Width of a video frame in pixels
+   * 
+   * @return int 
+   */
+  int Width() const {
+    return av_state_->codec_params_->width;
+  }
+
+  /**
+   * @brief Height of a video frame in pixels
+   * 
+   * @return int 
+   */
+  int Height() const {
+    return av_state_->codec_params_->height;
+  }
+
+  /**
+   * @brief Number of channels in a video
+   * 
+   * @return int 
+   */
+  int Channels() const {
+      return channels_;
+  }
+
+  /**
+   * @brief Total number of values in a frame (width * height * channels)
+   * 
+   * @return int 
+   */
+  int FrameSize() const {
+    return Channels() * Width() * Height();
+  }
+
+  /**
+   * @brief Reads next frame of the video and copies it to the provided buffer, if copy_to_output is True.
+   * 
+   * @param data Output buffer to copy data to.
+   * @param copy_to_output Whether copy data to output. 
+   */
+  void ReadNextFrame(uint8_t *data, bool copy_to_output = true);
```

Comment:
So, the metadata like pts, or the frame number are not important to us, and the frame number can be reliably tracked from the outside of this utility?
There is no chance of decoder giving us frames out of order or any of that? Or I'm just mixing it with B-frames, that it can be encoded with dependency into future, but not returned in that way?

Reply:
At this point there is no chance.
When we support more convoluted cases like you describe, I will add necessary mechanisms.